File: model/src/main/kotlin/net/dungeonhub/entity/model/InitializeModel.java
package net.dungeonhub.entity.model;

import me.taubsie.dungeonhub.common.entity.Entity;
import me.taubsie.dungeonhub.common.entity.InitializingFactory;

public interface InitializeModel<E extends Entity, C extends CreationModel> extends Model {
    E toEntity();

    InitializeModel<E, C> fromCreationModel(C creationModel);

    default E toEntity(InitializingFactory<E> factory) {
        return factory.transform(toEntity());
    }
}

File: model/src/main/kotlin/net/dungeonhub/structure/model/InitializeModel.kt
package net.dungeonhub.structure.model

import net.dungeonhub.structure.entity.Entity

interface InitializeModel<E : Entity<M>, M : Model, C : CreationModel> : Model {
    fun toEntity(): E

    fun fromCreationModel(creationModel: C): InitializeModel<E, M, C>
}